Add CarPlay storage API to iOS integration

bgoncal opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Added CarPlay storage to iOS integration and API's to update it's content. This will be used to define server global configurations, such as if CarPlay should be disabled for all and default entities to show in quick access tab

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

bgoncal committed:
Add CarPlay configuration to mobile_app

- Add carplay configuration schema with enabled and quick_access fields
- Support array of objects with entity_id and optional display_name
- Add comprehensive test coverage for configuration validation
- Update constants to avoid hardcoded strings
- Add example configuration file

bgoncal committed:
feat(ios): include carplay config from mobile_app integration

- Add carplay configuration to /api/ios/config endpoint
- Include carplay settings from mobile_app integration when available
- Add comprehensive test coverage for carplay config inclusion
- Handle cases where mobile_app integration is not loaded
- Maintain backward compatibility

bgoncal committed:
refactor: move carplay configuration from mobile_app to iOS integration

- Move carplay schema, constants, and configuration logic to iOS integration
- Add comprehensive carplay configuration validation and testing to iOS
- Update /api/ios/config endpoint to include local carplay configuration
- Remove all carplay-related code from mobile_app integration
- Add 6 new carplay tests to iOS integration with full validation coverage
- Maintain backward compatibility for iOS configuration
- Clean up unused imports in both integrations

The carplay configuration is now fully managed by the iOS integration:
- ios:
    carplay:
      enabled: true  # defaults to true
      quick_access:
        - entity_id: 'light.living_room'
          display_name: 'Living Room Light'  # optional
        - entity_id: 'climate.thermostat'     # display_name optional

bgoncal committed:
revert: restore mobile_app integration to original state

- Revert mobile_app/__init__.py and const.py to their original state
- Remove all carplay-related code from mobile_app integration
- All carplay functionality is now exclusively in iOS integration
- Mobile_app integration is completely untouched and clean

bgoncal committed:
final: ensure mobile_app integration is completely reverted

- Final cleanup to ensure mobile_app integration matches original state
- No carplay code remains in mobile_app integration
- All carplay functionality exclusively in iOS integration

bgoncal committed:
Using CarPlay storage instead of configuration.yaml

bgoncal committed:
Remove example_carplay_config.yaml as CarPlay now uses storage API

bgoncal committed:
fix: remove redundant checks and race conditions in CarPlay storage

- Remove unnecessary checks in async_setup_entry since CarPlay store is always initialized in async_setup first
- Remove redundant storage of CarPlay store in hass.data[DOMAIN] - only store in DATA_CARPLAY_STORAGE
- Add asyncio.Lock to prevent race conditions in async_load method
- Update tests to use proper CarPlay store access pattern
- Fix import organization and formatting issues

bgoncal commented on homeassistant/components/ios/__init__.py:
```diff
@@ -317,10 +334,20 @@ def __init__(self, config: dict[str, Any]) -> None:
+
+        # Add carplay config if store exists
+        if DOMAIN in hass.data and "carplay_store" in hass.data[DOMAIN]:
+            store = hass.data[DOMAIN]["carplay_store"]
```
Yeah sorry, I am pair programming with copilot since I don't know much about the codebase, it's not the smartest always :P

bgoncal commented:
@MartinHjelmare Whats the proper way that I can restrict this to Admin only?

bgoncal committed:
fix: typo in storage.py and improve test coverage to 85%

- Fix typo: 'retrn' -> 'return' in async_load method
- Add comprehensive tests for storage subscriptions and individual setters
- Add tests for iOS utility functions (devices_with_push, enabled_push_ids, etc.)
- Add tests for push config view and device identification view
- Add test for error handling when saving device data fails
- Total test coverage now at 85%, exceeding 84.92% target

MartinHjelmare commented:
I think we can use the `require_admin` decorator. Example:
https://github.com/home-assistant/core/blob/7ed8613411d1ca65fa6cc8ed2a8fab38e5258f43/homeassistant/components/config/view.py#L87-L98

MartinHjelmare commented on homeassistant/components/ios/storage.py:
```diff
@@ -0,0 +1,101 @@
+
+async def async_get_carplay_store(hass: HomeAssistant) -> CarPlayStore:
+    """Get the CarPlay store."""
+    if DATA_CARPLAY_STORAGE not in hass.data:
```
I dont think we should have this check here. Let's simplify things and create and load the store during setup in `async_setup`. Then we know that's done already in the view methods and we don't need a coroutine function to get the store. Ie we don't need to await anything to get a loaded store.

bgoncal committed:
refactor: address PR review feedback for CarPlay storage

- Remove asyncio.Lock from CarPlay storage (not needed since store is loaded during setup)
- Simplify store creation: load during async_setup instead of lazy loading
- Make get_carplay_store() synchronous since store is guaranteed to be loaded
- Update all view methods to use synchronous store access
- Fix import organization and test updates

Addresses reviewer feedback from MartinHjelmare in PR #157782

bgoncal committed:
Merge branch 'dev' into carplay-mobile-app-config

MartinHjelmare commented on homeassistant/components/ios/storage.py:
```diff
@@ -25,28 +24,22 @@ def __init__(self, hass: HomeAssistant) -> None:
 
     async def async_load(self) -> None:
         """Load the data from storage."""
         if self._loaded:
```
I'd remove this check and attribute too. We should only load the store once during the Home Assistant application lifetime in the current design.

bgoncal committed:
refactor: address remaining PR feedback

- Remove _loaded attribute and check from CarPlayStore as requested
- Simplify store to load only once during setup without redundant checks
- Refactor tests to focus on HTTP API rather than store implementation details
- Use hass_storage fixture for testing storage persistence
- Remove direct store operation tests in favor of API-driven tests

bgoncal committed:
fix: use integration setup and hass_storage fixture in test

- Replace manual store creation with proper integration setup
- Use hass_storage fixture to pre-populate test data instead of manual store operations
- Addresses remaining PR feedback about proper test patterns

bgoncal committed:
Merge branch 'dev' into carplay-mobile-app-config

MartinHjelmare commented on homeassistant/components/ios/storage.py:
```diff
@@ -0,0 +1,81 @@
+        self, on_update_callback: Callable[[], None]
+    ) -> Callable[[], None]:
+        """Subscribe to data changes."""
+        self.subscriptions.setdefault(None, []).append(on_update_callback)
```
Why do we need a dict if there's only one item with key `None`? We could just have a list of callbacks instead.

MartinHjelmare commented on homeassistant/components/ios/storage.py:
```diff
@@ -0,0 +1,81 @@
+        """Get the carplay data."""
+        return self.data.copy()
+
+    async def async_set_enabled(self, enabled: bool) -> None:
```
`async_set_enabled` is never used.

MartinHjelmare commented on homeassistant/components/ios/storage.py:
```diff
@@ -0,0 +1,81 @@
+        self.data["enabled"] = enabled
+        await self.async_save()
+
+    async def async_set_quick_access(self, quick_access: list[dict[str, Any]]) -> None:
```
`async_set_quick_access` is never used.

MartinHjelmare commented on homeassistant/components/ios/storage.py:
```diff
@@ -0,0 +1,81 @@
+            self.data = stored_data
+        else:
+            # Default carplay configuration
+            self.data = {
```
I'd use a dataclass instead so we can have typed easily accessible attributes.

Alternatively a typed dict.

https://docs.python.org/3/library/dataclasses.html

MartinHjelmare commented on homeassistant/components/ios/__init__.py:
```diff
@@ -358,3 +379,85 @@ async def post(self, request: web.Request) -> web.Response:
+            return self.json_message("Data must be an object", HTTPStatus.BAD_REQUEST)
+
+        # Validate enabled field if present
+        if "enabled" in data and not isinstance(data["enabled"], bool):
```
We can use the `RequestDataValidator` decorator with a voluptuous schema for validating the request data instead.

https://github.com/home-assistant/core/blob/8a10638470643e37a69a30aeeafe3d7d929ee023/homeassistant/components/http/data_validator.py#L21-L28

Example:
https://github.com/home-assistant/core/blob/8a10638470643e37a69a30aeeafe3d7d929ee023/homeassistant/components/backup/onboarding.py#L93-L109

MartinHjelmare commented on tests/components/ios/test_init.py:
```diff
@@ -59,3 +66,390 @@ async def test_not_configuring_ios_not_creates_entry(hass: HomeAssistant) -> Non
+
+    # Set up iOS config
+    ios_config = {"push": {"categories": []}}
+    view = iOSConfigView(ios_config)
```
We shouldn't test the view directly. We should always use the client to make request to the registered endpoints when testing the view.

MartinHjelmare commented on tests/components/ios/test_init.py:
```diff
@@ -59,3 +66,390 @@ async def test_not_configuring_ios_not_creates_entry(hass: HomeAssistant) -> Non
+    assert data == {"enabled": True, "quick_access": []}
+
+
+async def test_carplay_api_endpoints(
```
We're doing the same tests in `test_carplay_api_with_storage_data`.

Maybe parametrize that test instead with different initial stored data?

MartinHjelmare commented on tests/components/ios/test_init.py:
```diff
@@ -59,3 +66,390 @@ async def test_not_configuring_ios_not_creates_entry(hass: HomeAssistant) -> Non
+    assert hass_storage["ios.carplay_config"]["data"] == update_data
+
+
+async def test_carplay_api_error_handling(
```
I'd merge this test with `test_carplay_api_validation`.

MartinHjelmare commented on tests/components/ios/test_init.py:
```diff
@@ -59,3 +66,390 @@ async def test_not_configuring_ios_not_creates_entry(hass: HomeAssistant) -> Non
+    assert resp.status == 400
+
+
+async def test_ios_utility_functions(hass: HomeAssistant) -> None:
```
The following tests are not related to the PR subject and should preferably be moved to a separate PR.

bgoncal replied:
Removed

bgoncal committed:
refactor: address PR feedback with dataclass and clean storage

- Remove accidentally committed storage file
- Use dataclass for type-safe configuration structure
- Remove unused subscription and individual setter methods
- Make attributes protected (_hass, _data)
- Simplify get_data to synchronous operation
- Remove unnecessary copy() since data gets serialized

bgoncal committed:
Address PR feedback: Clean up tests and remove storage file

- Remove CarPlay storage test (test through HTTP API instead)
- Remove non-CarPlay related tests as suggested
- Clean up unused imports
- Remove accidentally committed storage file
- Keep only CarPlay-specific tests for this PR

bgoncal committed:
Apply suggestion from @MartinHjelmare

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bgoncal committed:
Remove non-CarPlay tests and simplify to 2 focused tests

- Remove tests for general iOS setup functionality
- Keep only test_carplay_get_data and test_carplay_set_data
- Focus tests on core CarPlay API functionality (GET/SET)
- Clean up test file to remove unrelated test cases
